packages/dd-trace/src/exporters/agent/writer.js

```diff
@@ -106,6 +100,41 @@ class Writer {
       headers[key] = value
     }
   }
+
+  _encode (trace) {
+    const offset = this._offset
+
+    try {
+      this._offset = encode(this._buffer, this._offset, trace)
+      this._count++
+    } catch (e) {
+      if (e.name.startsWith('RangeError')) {
+        if (offset === 0) {
+          return log.error('Dropping trace because its payload is too large.')
+        }
+
+        this._offset = offset
+
+        this.flush()
+        this._encode(trace)
+      } else {
+        log.error(e)
+      }
+
+      return
+    }
+
+    log.debug(() => [
+      'Added encoded trace to buffer:',
+      this._buffer.slice(offset, this._offset).toString('hex').match(/../g).join(' ')
+    ].join(' '))
+  }
+
+  _reset () {
+    this._buffer = Buffer.allocUnsafe(10 * 1024 * 1024) // 10mb
```

Comment:
> It should be less than 10mb since the agent allows 10mb max including other things. A safe value would be 8mb.

8MB is the soft limit. See https://github.com/DataDog/dd-trace-js/pull/858/files/7f490b33356fffdd28e5bc7f69c525e7d311ed92#diff-aa97d0a40b15e6cb805331a6bc6d6936R28

I set the buffer to be 2MB bigger than the soft limit so that if a trace comes in that crosses that boundary, we'll still let it through so long as it's smaller than 2MB.

> Also, why not just do this once per instance? Just clearing the first 5 bytes and updating the offset should be enough. It doesn't matter if garbage is still in the buffer.

Good point! Since writes are synchronous, we can do that. Also I don't even think the first 5 bytes need to be cleared, since we're just setting them afterward anyway, along with the offset.

Reply:
> I set the buffer to be 2MB bigger than the soft limit so that if a trace comes in that crosses that boundary, we'll still let it through so long as it's smaller than 2MB.

Why even let it through if we're going to drop it? The 8MB should be a hard limit to protect the agent.

Reply:
With the change above, we'll put in the 8MB hard limit.